src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1500,14 +1530,225 @@ public Durabilty getDurabilty() {
+            case SAVE:
+                return new Translog.Index();
+            default:
+                throw new IOException("No type for [" + type + "]");
+        }
+    }
+
+
+    @Override
+    public void prepareCommit() throws IOException {
+        ensureOpen();
+        TranslogWriter writer = null;
+        try (ReleasableLock lock = writeLock.acquire()) {
+            if (currentCommittingTranslog != null) {
+                throw new IllegalStateException("already committing a translog with generation: " + currentCommittingTranslog.getGeneration());
+            }
+            writer = current;
+            writer.sync();
+            currentCommittingTranslog = current.immutableReader();
+            Path checkpoint = location.resolve(CHECKPOINT_FILE_NAME);
+            assert Checkpoint.read(checkpoint).generation == currentCommittingTranslog.getGeneration();
+            Path commitCheckpoint = location.resolve(getCommitFileName(currentCommittingTranslog.getGeneration()));
+            Files.copy(checkpoint, commitCheckpoint);
+            IOUtils.fsync(commitCheckpoint, false);
+            IOUtils.fsync(commitCheckpoint.getParent(), true);
+            // create a new translog file - this will sync it and update the checkpoint data;
+            final TranslogWriter newFile = createWriter(current.getGeneration() + 1);
+            current = newFile;
+            // notify all outstanding views of the new translog (no views are created now as
+            // we hold a write lock).
+            for (FsView view : outstandingViews) {
+                view.onNewTranslog(currentCommittingTranslog.clone(), newFile.newReaderFromWriter());
+            }
+            IOUtils.close(writer);
+            logger.trace("current translog set to [{}]", current.getGeneration());
+            assert writer.syncNeeded() == false : "old translog writer must not need a sync";
+
+        } catch (Throwable t) {
+            close(); // tragic event
+            throw t;
+        }
+    }
+
+    @Override
+    public void commit() throws IOException {
+        ensureOpen();
+        ImmutableTranslogReader toClose = null;
+        try (ReleasableLock lock = writeLock.acquire()) {
+            if (currentCommittingTranslog == null) {
+                prepareCommit();
+            }
+            current.sync();
+            lastCommittedTranslogFileGeneration = current.getGeneration(); // this is important - otherwise old files will not be cleaned up
+            if (recoveredTranslogs.isEmpty() == false) {
+                IOUtils.close(recoveredTranslogs);
+                recoveredTranslogs.clear();
             }
+            toClose = this.currentCommittingTranslog;
+            this.currentCommittingTranslog = null;
+        } finally {
+            IOUtils.close(toClose);
+        }
+    }
+
+    @Override
+    public void rollback() throws IOException {
+        ensureOpen();
+        close();
+    }
```

Comment:
Hmm, why do we need to sync `current` here?  Isn't current the "next" translog file, i.e. `prepareCommit` set up currentCommittingTranslog (= current-1) for committing?